Add support for context

suhaibmujahid opened the pull request:
# PR Description
This PR is in response to the changes requested by @ghostsquad in #182.

_What does this fix or add?_
It adds support for `context` in all of the API requests.

# Checklist

* [x] Tests added
  * [x] Good Path
  * [x] Error Path
* [x] Commits follow conventions described here:
  * [x] [https://conventionalcommits.org/en/v1.0.0-beta.4/#summary](https://conventionalcommits.org/en/v1.0.0-beta.4/#summary)
  * [x] [https://chris.beams.io/posts/git-commit/#seven-rules](https://chris.beams.io/posts/git-commit/#seven-rules)
* [x] Commits are squashed such that
  * [x] There is 1 commit per isolated change
* [x] I've not made extraneous commits/changes that are unrelated to my change.

ghostsquad commented:
@suhaibmujahid this is great! thank you. I'll be reviewing this over the weekend.

suhaibmujahid commented:
@ghostsquad great, please let me know if you want me to perform any changes.

andygrunwald commented on issue.go:
```diff
@@ -1124,6 +1225,11 @@ func (s *IssueService) DoTransitionWithPayload(ticketID, payload interface{}) (*
 	return resp, err
 }
 
+// XXXXX wraps XXXXXWithContext using the background context.
```
```suggestion
// DoTransitionWithPayload wraps DoTransitionWithPayloadWithContext using the background context.
```

andygrunwald commented on jira.go:
```diff
@@ -108,7 +109,7 @@ func NewClient(httpClient httpClient, baseURL string) (*Client, error) {
 // NewRawRequest creates an API request.
```
```suggestion
// NewRawRequestWithContext creates an API request.
```

andygrunwald commented on issuelinktype.go:
```diff
@@ -79,12 +90,17 @@ func (s *IssueLinkTypeService) Create(linkType *IssueLinkType) (*IssueLinkType,
+	return s.CreateWithContext(context.Background(), linkType)
+}
+
 // Update updates an issue link type.  The issue is found by key.
```
```suggestion
// UpdateWithContext updates an issue link type.  The issue is found by key.
```

andygrunwald commented on issuelinktype.go:
```diff
@@ -96,16 +112,26 @@ func (s *IssueLinkTypeService) Update(linkType *IssueLinkType) (*IssueLinkType,
 //
 // JIRA API docs: https://developer.atlassian.com/cloud/jira/platform/rest/v2/#api-rest-api-2-issueLinkType-issueLinkTypeId-delete
-func (s *IssueLinkTypeService) Delete(ID string) (*Response, error) {
+func (s *IssueLinkTypeService) linkType(ctx context.Context, ID string) (*Response, error) {
```
This seem to by a typo.
Is this ... ?

```suggestion
func (s *IssueLinkTypeService) DeleteWithContext(ctx context.Context, ID string) (*Response, error) {
```

andygrunwald commented on request_legacy.go:
```diff
@@ -0,0 +1,18 @@
+// +build !go1.13
```
It would be a good idea to have a comment here and link what has changed in go v1.13 and why it is necessary.

suhaibmujahid replied:
I totally agree with you. It is important for future maintainers. I will do it.

andygrunwald commented on request_context.go:
```diff
@@ -0,0 +1,13 @@
+// +build go1.13
```
It would be a good idea to have a comment here and link what has changed in go v1.13 and why it is necessary.

suhaibmujahid replied:
I totally agree with you. It is important for future maintainers. I will do it.

andygrunwald commented on metaissue.go:
```diff
@@ -43,15 +44,22 @@ type MetaIssueType struct {
 
 // GetCreateMetaWithOptions makes the api call to get the meta information without requiring to have a projectKey
-func (s *IssueService) GetCreateMetaWithOptions(options *GetQueryOptions) (*CreateMetaInfo, *Response, error) {
+// todo: the name should be reconsidered, GetCreateMetaWithOptions could be replace with simply
```
Can you open a ticket for this instead of having a TODO in the code?

suhaibmujahid replied:
Done! The issue number is #294.

suhaibmujahid commented:
Thank you @andygrunwald for the code review. 

I did rebase based on the master, solve the conflicts, fixed the comments for Godocs, and incorporated your other suggestions.

 For unit tests, I do not think we need to do anything for now for the following reasons:
1. All the <NAME>WithContext functions are implicit tested by the original unit tests as you mentioned.
2. The functions will be renamed in `v2` to remove the `WithContext` suffixes as proposed originally by @ghostsquad (I suggest to keep `NewRequestWithContext` since it match the standard `http` package).
3. I do not think the new tests will add any value if we just copy the old tests and call them with `context.Background()` since this is what already happening now.

Thanks again!

andygrunwald commented:
Thanks @suhaibmujahid.
Looks good to me. I will go ahead and merge it.

From now, I don't see anything that we will break. If we break something, we will either fix it and move forward or roll it back and see what happens. In the end, it is still a huge PR.

Thanks a lot for your effort here.
I would be curious to get your experience from your application use case, using the new lib with context support.
